src/persistent_tasks.jl

```diff
@@ -0,0 +1,156 @@
+and finishing precompilation exceeds time `tmax`.
+
+# How to fix failing packages
+
+Often, the easiest fix is to modify the `__init__` function to check whether the
+Julia process is precompiling some other package; if so, don't launch the
+persistent `Task`s.
+
+```
+function __init__()
+    # Other setup code here
+    if ccall(:jl_generating_output, Cint, ()) == 0   # if we're not precompiling...
+        # launch persistent tasks here
+    end
+end
+```
+
+In more complex cases, you may need to set up independently-callable functions
+to launch the tasks and cleanly shut them down.
+
+# Arguments
+- `package`: a top-level `Module` or `Base.PkgId`.
+
+# Keyword Arguments
+- `tmax::Real`: the maximum time (in seconds) to wait after loading the
+  package before forcibly shutting down the precompilation process (triggering
+  a test failure).
+"""
+function test_persistent_tasks(package::PkgId; tmax = 10, fails::Bool = false)
+    @testset "$package persistent_tasks" begin
+        result = root_project_or_failed_lazytest(package)
+        result isa LazyTestResult && return result
+        @test fails ⊻ precompile_wrapper(result, tmax)
+    end
+end
+
+function test_persistent_tasks(package::Module; kwargs...)
+    test_persistent_tasks(PkgId(package); kwargs...)
+end
+
+"""
+    Aqua.test_persistent_tasks_deps(package; fails = Dict{String,Bool}(), kwargs...)
+
+Test all the dependencies of `package` with [`Aqua.test_persistent_tasks`](@ref).
+`get(fails, dep, false)` encodes whether dependency `dep` is expected to fail the test
+(this is primarily intended for use in Aqua's own internal test suite).
+Any additional kwargs (e.g., `tmax`) are passed to [`Aqua.test_persistent_tasks`](@ref).
+"""
+function test_persistent_tasks_deps(package::PkgId; fails = Dict{String,Bool}(), kwargs...)
+    result = root_project_or_failed_lazytest(package)
+    result isa LazyTestResult && return result
+    prj = TOML.parsefile(result)
+    deps = get(prj, "deps", nothing)
+    @testset "$result" begin
+        if deps === nothing
+            return LazyTestResult("$package", "`$result` does not have `deps`", true)
+        else
+            for (name, uuid) in deps
+                id = PkgId(UUID(uuid), name)
+                test_persistent_tasks(id; fails = get(fails, name, false), kwargs...)
+            end
+        end
+    end
+end
+
+function test_persistent_tasks_deps(package::Module; kwargs...)
+    test_persistent_tasks_deps(PkgId(package); kwargs...)
+end
+
+function precompile_wrapper(project, tmax)
+    pkgdir = dirname(project)
+    pkgname = basename(pkgdir)
+    wrapperdir = tempname()
+    wrappername, _ = only(Pkg.generate(wrapperdir))
+    Pkg.activate(wrapperdir)
+    Pkg.develop(PackageSpec(path = pkgdir))
+    statusfile = joinpath(wrapperdir, "done.log")
+    open(joinpath(wrapperdir, "src", wrappername * ".jl"), "w") do io
+        println(
+            io,
+            """
+module $wrappername
```

Comment:
This can be indented.

Reply:
It was originally, but this package requires a JuliaFormatter test and it won't let me 🤣.